IPsec unable to use 36 byte psk for GCM-256-AES

### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We use Cilium IPsec transparant encryption.  When attempting to upgrade a cluster to Cilium 1.15.6 and enabling the per-tunnel keys Cilium starts reporting errors when loading the new key.

### Cilium Version

1.15.6

### Kernel Version

5.15.156-102.160.amzn2.x86_64

### Kubernetes Version

EKS 1.27

### Regression

yes.  We haven't done a bisection but we previously ran 1.14.3.

### Sysdump

_No response_

### Relevant log output

```shell
panic: runtime error: slice bounds out of range [:36] with length 32

goroutine 333 [running]:
github.com/cilium/cilium/pkg/datapath/linux/ipsec.computeNodeIPsecKey({0xc0085a2640, 0x24, 0x3035312e37332e?}, {0xc0052c1668, 0x4, 0x7feb0054b330?}, {0xc008ec2bfc, 0x4, 0x43ad31?}, {0xc009562600, ...}, ...)
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/ipsec/ipsec_linux.go:182 +0x2e5
github.com/cilium/cilium/pkg/datapath/linux/ipsec.deriveNodeIPsecKey(0xc00a4108d0, {0xc0052c1668, 0x4, 0x198}, {0xc008ec2bf0, 0x10, 0x10}, {0xc0005d2b10, 0x24}, {0xc002020e40, ...})
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/ipsec/ipsec_linux.go:209 +0x252
github.com/cilium/cilium/pkg/datapath/linux/ipsec.getNodeIPsecKey({0xc0052c1668, 0x4, 0x198}, {0xc008ec2bf0, 0x10, 0x10}, {0xc0005d2b10, 0x24}, {0xc002020e40, 0x24}, ...)
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/ipsec/ipsec_linux.go:246 +0xda
github.com/cilium/cilium/pkg/datapath/linux/ipsec.ipSecReplaceStateOut({0xc0052c1668, 0x4, 0x198}, {0xc008ec2bf0, 0x10, 0x10}, 0x4242, {0xc0005d2b10?, 0xc008ec2be0?}, {0xc002020e40, ...}, ...)
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/ipsec/ipsec_linux.go:549 +0x88
github.com/cilium/cilium/pkg/datapath/linux/ipsec.UpsertIPsecEndpoint(0xc00a410900?, 0xc001434000?, {0xc0052c1668, 0x4, 0x198}, {0xc008ec2bf0, 0x10, 0x10}, 0xd450?, {0xc002020e40, ...}, ...)
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/ipsec/ipsec_linux.go:936 +0x48c
github.com/cilium/cilium/pkg/datapath/linux.(*linuxNodeHandler).enableIPsecIPv4(0xc000991380, 0xc00876dcc0, 0x2e3e?, 0x50?, 0x0?)
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/ipsec.go:216 +0x161b
github.com/cilium/cilium/pkg/datapath/linux.(*linuxNodeHandler).enableIPsec(0xc000991380, 0x0, 0xc00876dcc0, 0x5?)
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/ipsec.go:140 +0x2f7
github.com/cilium/cilium/pkg/datapath/linux.(*linuxNodeHandler).nodeUpdate(0xc000991380, 0x0, 0xc00876dcc0, 0x0)
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/node.go:945 +0xb98
github.com/cilium/cilium/pkg/datapath/linux.(*linuxNodeHandler).AllNodeValidateImplementation(0xc000991380)
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/node.go:1317 +0x14a
github.com/cilium/cilium/pkg/datapath/linux/ipsec.keyfileWatcher({0x3f6c2a0, 0xc00646e420}, 0xc00646e330, {0xc0011e5580, 0xf}, {0x3f79d80, 0xc000991380}, {0x3f56ef0, 0xc004c28e00})
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/ipsec/ipsec_linux.go:1258 +0x3b5
github.com/cilium/cilium/pkg/datapath/linux/ipsec.StartKeyfileWatcher.func1({0x3f6c2a0?, 0xc00646e420?}, {0x3f56ef0?, 0xc004c28e00?})
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/ipsec/ipsec_linux.go:1291 +0x47
github.com/cilium/cilium/pkg/hive/job.(*jobOneShot).start(0xc0064209c0, {0x3f6c2a0, 0xc00646e420}, 0xc005582fd0?, {0x3f574d0, 0xc0008508f8}, {{{0xc001823ba0, 0x1, 0x1}}, {0x3f9e8f0, ...}, ...})
	/go/src/github.com/cilium/cilium/pkg/hive/job/job.go:347 +0x6b2
created by github.com/cilium/cilium/pkg/hive/job.(*group).add.func1 in goroutine 1
	/go/src/github.com/cilium/cilium/pkg/hive/job/job.go:210 +0x18d
```


### Anything else?

This seems like a byproduct of using [sha256](https://github.com/cilium/cilium/blob/5614531067a83e20d24bccc7b12b314330d043c3/pkg/datapath/linux/ipsec/ipsec_linux.go#L188), which yields a 32 byte result.

For GCM-256-AES, the PSK needs to be 36 bytes.

Ref: https://datatracker.ietf.org/doc/rfc4106/#

```
AES-GCM-ESP with a 256 bit key
      The KEYMAT requested for each AES GCM key is 36 octets.  The first
      32 octets are the 256-bit AES key, and the remaining four octets
      are used as the salt value in the nonce.
```

I think switching to SHA-512 may help?


### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/pkg/datapath/linux/ipsec/ipsec_linux.go
+++ b/pkg/datapath/linux/ipsec/ipsec_linux.go
@@ -10,6 +10,7 @@ import (
 	"bytes"
 	"context"
 	"crypto/sha256"
+	"crypto/sha512"
 	"encoding/hex"
 	"errors"
 	"fmt"
@@ -184,8 +185,16 @@ func computeNodeIPsecKey(globalKey, srcNodeIP, dstNodeIP, srcBootID, dstBootID [
 	input = append(input, dstNodeIP...)
 	input = append(input, srcBootID[:36]...)
 	input = append(input, dstBootID[:36]...)
-	output := sha256.Sum256(input)
-	return output[:len(globalKey)]
+
+	var hash []byte
+	if len(globalKey) <= 32 {
+		h := sha256.Sum256(input)
+		hash = h[:]
+	} else {
+		h := sha512.Sum512(input)
+		hash = h[:]
+	}
+	return hash[:len(globalKey)]
 }
 
 // canonicalIP returns a canonical IPv4 address (4 bytes)

--- a/pkg/datapath/linux/ipsec/ipsec_linux_test.go
+++ b/pkg/datapath/linux/ipsec/ipsec_linux_test.go
@@ -42,6 +42,7 @@ var (
 	path           = "ipsec_keys_test"
 	keysDat        = []byte("1 hmac(sha256) 0123456789abcdef0123456789abcdef cbc(aes) 0123456789abcdef0123456789abcdef\n1 hmac(sha256) 0123456789abcdef0123456789abcdef cbc(aes) 0123456789abcdef0123456789abcdef foobar\n1 digest_null \"\" cipher_null \"\"\n")
 	keysAeadDat    = []byte("6 rfc4106(gcm(aes)) 44434241343332312423222114131211f4f3f2f1 128\n")
+	keysAeadDat256 = []byte("6 rfc4106(gcm(aes)) 44434241343332312423222114131211f4f3f2f144434241343332312423222114131211 128\n")
 	invalidKeysDat = []byte("1 test abcdefghijklmnopqrstuvwzyzABCDEF test abcdefghijklmnopqrstuvwzyzABCDEF\n")
 )
 
@@ -71,16 +72,14 @@ func TestInvalidLoadKeys(t *testing.T) {
 func TestLoadKeys(t *testing.T) {
 	log := setupIPSecSuitePrivileged(t)
 
-	keys := bytes.NewReader(keysDat)
-	_, spi, err := LoadIPSecKeys(log, keys)
-	require.NoError(t, err)
-	err = SetIPSecSPI(log, spi)
-	require.NoError(t, err)
-	keys = bytes.NewReader(keysAeadDat)
-	_, spi, err = LoadIPSecKeys(log, keys)
-	require.NoError(t, err)
-	err = SetIPSecSPI(log, spi)
-	require.NoError(t, err)
+	testCases := [][]byte{keysDat, keysAeadDat, keysAeadDat256}
+	for _, testCase := range testCases {
+		keys := bytes.NewReader(testCase)
+		_, spi, err := LoadIPSecKeys(log, keys)
+		require.NoError(t, err)
+		err = SetIPSecSPI(log, spi)
+		require.NoError(t, err)
+	}
 }
 
 func TestParseSPI(t *testing.T) {

